File: api/Validators/invoice.js
'use strict';

/**
 * api/Validators/invoice.js
 *
 * Joi schemas for the invoices resource (sales + purchase vouchers). Mirrors the
 * `invoices` (migration 20260101000016) and `invoice_items` (20260101000017)
 * tables — but only the fields a client may set. Server-managed columns
 * (id, company_id, invoice_no, all money totals, cgst/sgst/igst, round_off,
 * tally_*, created_by, timestamps, deleted_at) are intentionally NOT accepted
 * from the body; they are computed / stamped by the InvoiceController inside the
 * write transaction. NEVER trust client-sent totals.
 *
 * Schemas:
 *   createSalesInvoiceSchema    — POST /invoices/sales    (customer_id required)
 *   createPurchaseInvoiceSchema — POST /invoices/purchase (supplier_id required,
 *                                                          + supplier_bill_no)
 *   listInvoiceSchema           — GET  /invoices/...       (pagination + filters)
 *
 * Conventions:
 *   • FK ids are positive integers; existence is enforced by the DB FK, not here.
 *   • blank optional strings are allowed via `.allow('', null)` so a client can
 *     omit / clear a field.
 *   • items is an array with at least one line; per-line quantity > 0 and
 *     rate >= 0 are required, discount_pct (0..100) and gst_rate default to 0.
 *   • `status` walks the Tally sync lifecycle and defaults to 'pending_tally'.
 */

const Joi = require('joi');

// Invoice sync-lifecycle states — matches invoices.status default.
const STATUSES = ['pending_tally', 'sent_to_tally', 'created', 'failed'];

// Reusable optional positive-integer FK.
const fkId = Joi.number().integer().positive();

// Reusable optional short/long text — trimmed, blank/null allowed.
const optText = (max) => Joi.string().trim().max(max).allow('', null);

/**
 * A single invoice line. The money columns (taxable / gst_amount / amount) are
 * NOT accepted here — the controller computes them authoritatively from
 * quantity / rate / discount_pct / gst_rate.
 */
const itemSchema = Joi.object({
    product_id:   fkId.allow(null),
    description:  optText(2000),
    hsn:          optText(20),
    quantity:     Joi.number().greater(0).precision(2).required().messages({
        'number.base':     'Item quantity is required.',
        'number.greater':  'Item quantity must be greater than 0.',
        'any.required':    'Item quantity is required.',
    }),
    unit:         optText(30),
    rate:         Joi.number().min(0).precision(2).required().messages({
        'number.base':  'Item rate is required.',
        'number.min':   'Item rate must be 0 or more.',
        'any.required': 'Item rate is required.',
    }),
    discount_pct: Joi.number().min(0).max(100).precision(2).default(0),
    gst_rate:     Joi.number().min(0).precision(2).default(0),
});

const itemsArray = Joi.array().items(itemSchema).min(1).required().messages({
    'array.min':    'At least one line item is required.',
    'array.base':   'At least one line item is required.',
    'any.required': 'At least one line item is required.',
});

/**
 * POST /api/v1/invoices/sales
 * `customer_id`, `invoice_date` and at least one item are required.
 */
const createSalesInvoiceSchema = Joi.object({
    customer_id:     fkId.required().messages({
        'any.required':    'Customer is required.',
        'number.base':     'Customer is required.',
        'number.positive': 'Customer is required.',
    }),
    location_id:     fkId,
    sales_person_id: fkId,

    invoice_date:    Joi.date().iso().required().messages({
        'date.base':    'Invoice date is required.',
        'date.format':  'Invoice date must be an ISO date.',
        'any.required': 'Invoice date is required.',
    }),
    due_date:        Joi.date().iso().allow(null),
    notes:           optText(2000),

    status:          Joi.string().valid(...STATUSES).default('pending_tally'),

    items:           itemsArray,
});

/**
 * POST /api/v1/invoices/purchase
 * Same shape as sales but keyed on `supplier_id` (required) instead of
 * customer_id, carries an optional `supplier_bill_no`, and has no
 * sales_person_id.
 */
const createPurchaseInvoiceSchema = Joi.object({
    supplier_id:      fkId.required().messages({
        'any.required':    'Supplier is required.',
        'number.base':     'Supplier is required.',
        'number.positive': 'Supplier is required.',
    }),
    location_id:      fkId,
    supplier_bill_no: optText(60),

    invoice_date:     Joi.date().iso().required().messages({
        'date.base':    'Invoice date is required.',
        'date.format':  'Invoice date must be an ISO date.',
        'any.required': 'Invoice date is required.',
    }),
    due_date:         Joi.date().iso().allow(null),
    notes:            optText(2000),

    status:           Joi.string().valid(...STATUSES).default('pending_tally'),

    items:            itemsArray,
});

/**
 * GET /api/v1/invoices/{sales,purchase} (query string)
 * Pagination + the filters the list handlers read. customer_id / supplier_id
 * narrow to a party; date_from / date_to bound invoice_date.
 */
const listInvoiceSchema = Joi.object({
    search:      Joi.string().trim().max(191).allow('', null),
    status:      Joi.string().valid(...STATUSES),
    page:        Joi.number().integer().min(1).default(1),
    per_page:    Joi.number().integer().min(1).max(100).default(10),
    customer_id: fkId,
    supplier_id: fkId,
    date_from:   Joi.date().iso(),
    date_to:     Joi.date().iso(),
});

module.exports = {
    createSalesInvoiceSchema,
    createPurchaseInvoiceSchema,
    listInvoiceSchema,
    STATUSES,
};

File: app/lib/data/models/invoice.dart
/// One invoice header from `GET /api/v1/sales-invoices` (or purchase). Node
/// left-joins customer / supplier / location names. The money columns are
/// computed server-side (never trust client totals). pg returns numeric/bigint
/// columns as strings, so coercions are defensive.
class Invoice {
  const Invoice({
    required this.id,
    required this.invoiceNo,
    this.type,
    this.customer,
    this.supplier,
    this.location,
    this.invoiceDate,
    this.dueDate,
    this.subtotal,
    this.discount,
    this.taxable,
    this.taxAmount,
    this.roundOff,
    this.total,
    this.status,
    this.notes,
    this.items = const [],
  });

  final int id;
  final String invoiceNo;
  final String? type; // sales | purchase
  final String? customer;
  final String? supplier;
  final String? location;
  final String? invoiceDate;
  final String? dueDate;
  final num? subtotal;
  final num? discount;
  final num? taxable;
  final num? taxAmount;
  final num? roundOff;
  final num? total;
  final String? status; // pending_tally | sent_to_tally | created | failed
  final String? notes;
  final List<InvoiceItem> items;

  /// The party label that applies to this voucher kind (customer for sales,
  /// supplier for purchase) — whichever the join filled in.
  String? get party => customer ?? supplier;

  factory Invoice.fromJson(Map<String, dynamic> j) {
    final rawItems = (j['items'] is List) ? j['items'] as List : const [];
    return Invoice(
      id: _toInt(j['id']) ?? 0,
      invoiceNo: _s(j['invoice_no']),
      type: _sn(j['type']),
      customer: _sn(j['customer']),
      supplier: _sn(j['supplier']),
      location: _sn(j['location']),
      invoiceDate: _sn(j['invoice_date']),
      dueDate: _sn(j['due_date']),
      subtotal: _toNum(j['subtotal']),
      discount: _toNum(j['discount']),
      taxable: _toNum(j['taxable']),
      taxAmount: _toNum(j['tax_amount']),
      roundOff: _toNum(j['round_off']),
      total: _toNum(j['total']),
      status: _sn(j['status']),
      notes: _sn(j['notes']),
      items: rawItems
          .whereType<Map>()
          .map((m) => InvoiceItem.fromJson(m.cast<String, dynamic>()))
          .toList(growable: false),
    );
  }
}

/// One invoice line from `invoice_items` (returned nested on GET :id). The
/// taxable / gst_amount / amount are server-computed.
class InvoiceItem {
  const InvoiceItem({
    this.productId,
    this.description,
    this.hsn,
    this.quantity,
    this.unit,
    this.rate,
    this.discountPct,
    this.taxable,
    this.gstRate,
    this.gstAmount,
    this.amount,
  });

  final int? productId;
  final String? description;
  final String? hsn;
  final num? quantity;
  final String? unit;
  final num? rate;
  final num? discountPct;
  final num? taxable;
  final num? gstRate;
  final num? gstAmount;
  final num? amount;

  factory InvoiceItem.fromJson(Map<String, dynamic> j) => InvoiceItem(
        productId: _toInt(j['product_id']),
        description: _sn(j['description']),
        hsn: _sn(j['hsn']),
        quantity: _toNum(j['quantity']),
        unit: _sn(j['unit']),
        rate: _toNum(j['rate']),
        discountPct: _toNum(j['discount_pct']),
        taxable: _toNum(j['taxable']),
        gstRate: _toNum(j['gst_rate']),
        gstAmount: _toNum(j['gst_amount']),
        amount: _toNum(j['amount']),
      );
}

/// A short, friendly label for the invoice sync-lifecycle status. The raw
/// values (`pending_tally`, `sent_to_tally`, `created`, `failed`) map to words
/// that also colour correctly via `statusColor` (prefix families).
String invoiceStatusLabel(String? s) {
  switch ((s ?? '').toLowerCase()) {
    case 'pending_tally':
      return 'Pending';
    case 'sent_to_tally':
      return 'Sent';
    case 'created':
      return 'Synced';
    case 'failed':
      return 'Failed';
    default:
      return (s == null || s.isEmpty) ? 'Unknown' : s;
  }
}

String _s(Object? v) => v == null ? '' : v.toString();
String? _sn(Object? v) {
  if (v == null) return null;
  final s = v.toString().trim();
  return s.isEmpty ? null : s;
}

int? _toInt(Object? v) {
  if (v == null) return null;
  if (v is num) return v.toInt();
  final s = v.toString().trim();
  return s.isEmpty ? null : int.tryParse(s);
}

num? _toNum(Object? v) {
  if (v == null) return null;
  if (v is num) return v;
  final s = v.toString().trim();
  return s.isEmpty ? null : num.tryParse(s);
}
